core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1263,21 +1445,205 @@ class KafkaApis(val requestChannel: RequestChannel,
+        breakable{
+          topic.partitions().forEach ( partition => {
+            if (partition.acknowledgementBatches() != null && !partition.acknowledgementBatches().isEmpty) {
+              isAcknowledgeDataPresent = true
+              break()
+            } else {
+              isAcknowledgeDataPresent = false
+            }
+          })
+        }
+      })
+      isAcknowledgeDataPresent
+    }
+
+    var shareAcknowledgeResponse : ShareAcknowledgeResponse = null
+    var shareFetchResponse : ShareFetchResponse = null
+
+    // Handling the Acknowledgements from the ShareFetchRequest
+    val shareSessionError: Errors = sharePartitionManager.acknowledgeShareSessionCacheUpdate(groupId, Uuid.fromString(memberId), shareSessionEpoch)
+    if (shareSessionError.code() != Errors.NONE.code()) {
+      shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, shareSessionError.exception) match {
+        case response: ShareAcknowledgeResponse => response
+        case _ => null
+      }
+    } else if (!isAcknowledgeDataPresentInFetchRequest()) {
+      shareAcknowledgeResponse = ShareAcknowledgeResponse.of(
+        Errors.NONE,
+        AbstractResponse.DEFAULT_THROTTLE_TIME,
+        new util.LinkedHashMap[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData],
+        Collections.emptyList())
+    } else if (shareSessionEpoch == ShareFetchMetadata.INITIAL_EPOCH && isAcknowledgeDataPresentInFetchRequest()) {
+        // If this is a full fetch request, it should not contain any acknowledgement Data
+        shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception) match {
+          case response: ShareAcknowledgeResponse => response
+          case _ => null
+        }
+      }
+    else {
+      shareAcknowledgeResponse = handleAcknowledgeFromShareFetchRequest(request, topicNames, sharePartitionManager)
+    }
+
+    val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
+    val shareFetchContext = sharePartitionManager.newContext(groupId, shareFetchData, forgottenTopics, topicNames, newReqMetadata)
+
+    // Handling the Fetch from the ShareFetchRequest
+    // Fetching should not proceed under any one of the following scenarios -
+    // 1. If shareAcknowledgeResponse has Errors.INVALID_REQUEST error code
+    // 2. If this is a Final Fetch request (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) and shareFetchData is not empty
+    // 3. If this is a Final Fetch request (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) and forgottenTopics is not empty
+    if (shareAcknowledgeResponse.data().errorCode() ==  Errors.INVALID_REQUEST.code()
+      || (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH && (!shareFetchData.isEmpty || !forgottenTopics.isEmpty))) {
```

Comment:
In the case where it is FINAL_EPOCH, we do not want to fetch at all. We might have acknowledgements to process. It doesn't matter whether shareFetchData or forgottenTopics is empty. We process the acknowledgements, close the share session and do not fetch data.

Reply:
[This](https://confluentinc.atlassian.net/wiki/spaces/AK/pages/3334014224/KIP-932+Share+sessions) doc says that if session epoch is -1 (final fetch request), and if it contains topics to add or forget, the request should fail with the error INVALID_REQUEST

Reply:
By "topics to add", I mean if there are topics in the request which do not exist in the session, then fail it. But you can legitimately have entries in the topics array that are just acknowledgements for topic-partitions which are already in the session. Those should be processed, and then the session should be removed from the cache.